implement env variable expansion for kaniko builds

DanielSel opened the pull request:
Adds environment variable expansion for kaniko builds (to address currently inconsistent behaviour between buildArgs and envs).
See #3974

DanielSel commented:
@tejal29 @gsquared94 it seems like some of the unit tests are flaky in the CI, but run flawless locally. Took me a few attempts to get the build to pass: https://github.com/GoogleContainerTools/skaffold/runs/907521989

Now I removed the dummy commit via force-push and the build status went from passing to failing without a rebuild. Do you have any ideas what I should do?

DanielSel commented:
The Travis CI build is back to green. Thanks, in case anyone manually triggered the rebuild!

gsquared94 commented:
@DanielSel sorry about the long time to reply; was traveling. I'd restarted the failed job manually, travis can be flaky that way sometimes.

DanielSel commented:
I fixed the oversight in parsing image names (forgot the possibility of IP:PORT as registry url) and added a test.
@gsquared94 Could you re-run kokoro?

Thanks!

DanielSel commented:
@gsquared94 one last kokoro run please :)

nkubala commented on pkg/skaffold/build/cluster/kaniko.go:
```diff
@@ -110,3 +123,89 @@ func (b *Builder) copyKanikoBuildContext(ctx context.Context, workspace string,
+	return generatedEnvs, nil
+}
+
+func parseImageParts(imageStr string) (string, string, string, error) {
```
I think you should be able to use [`docker.ParseReference`](https://godoc.org/github.com/containers/image/docker#ParseReference) here instead. we do this a few other places in the code

DanielSel replied:
That's how I started out until I realized that `docker.ParseReference ` can't split out the image name from the repo path, which we need for our CI setup. 

I could replace part of the function code with a call to `docker.ParseReference` and only leave the code to split IMAGE_NAME from IMAGE_REPO if you prefer.

What do you think @nkubala ?

nkubala replied:
sorry I totally missed this one. I was actually wrong on my original comment, we're using docker's (`reference.Parse`)[https://godoc.org/github.com/docker/distribution/reference#Parse] in skaffold. we finesse that into a `ImageReference` struct which should be able to split image name and repo path (we call it `Domain`) right?

if it isn't doing everything you need, I think that code should live somewhere near here so we can embed those pieces into the `ImageReference` struct.

DanielSel replied:
@nkubala no worries. I'm not sure that it would make sense to refactor the ImageReference to change the way it splits image FQDN's since I assume there was a reason for this specific split.

However, I refactored my code to use docker.ParseReference (which returns the ImageReference you pointed out) instead of implementing it from scratch. Was that what you had in mind?

nkubala replied:
hey @DanielSel, I dug in a little more into `ParseReference` and how it creates the `ImageReference`. take the image `gcr.io/k8s-skaffold/skaffold:foo` as an example:

BaseName -> `gcr.io/k8s-skaffold/skaffold`
Domain -> `gcr.io`
Path -> `k8s-skaffold/skaffold`

the part you're interested in here is the "Repo", which would be `gcr.io/k8s-skaffold`. I think part of the confusion is that our terminology is not quite right here: `BaseName` is really "Unqualified Image Name" (to me at least), which is also "Repo" + "Image Name" (both of which are missing from this struct).

I propose that you take your logic here to split the image parts (which seems mostly fine to me as is) and move it into `docker.ParseReference`, and then embed "Repo" and "Image Name" into the `ImageReference` struct we already have. this won't change the existing usage of the struct around the codebase, but could be used later on if we wanted to, and is easier to find if we ever want to make changes to it. this isn't kaniko-specific logic, so it doesn't really belong in the kaniko package. WDYT?

DanielSel replied:
(I think) I got you now :)
Code updated.

DanielSel commented:
@tejal29 I took the freedom to rebase, I hope that's ok.

Could somebody trigger another kokoro run and make a decision about the requested changes please? This pull request is starting to gather dust :)

nkubala commented on pkg/skaffold/build/cluster/kaniko_test.go:
```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2019 The Skaffold Authors
```
```suggestion
Copyright 2020 The Skaffold Authors
```

nkubala reviewed (commented):
@DanielSel thank you so much for sticking with us here, this is very close but I just want to make sure we put the code in the right place the first time so we don't have to move it later :)

nkubala commented:
@DanielSel yep this is more what I was thinking! looks like the error you introduced in `ParseReference` is causing test failures though - maybe because this is being called before the default-repo is substituted on the image? I don't think we want to error here either way, probably just leave the field unpopulated. it should still give you the result you want when you're using `Repo`.

DanielSel commented:
@nkubala Sorry, my mistake. Fixed now.

tejal29 commented on pkg/skaffold/build/cluster/kaniko_test.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2019 The Skaffold Authors
+Copyright 2020 The Skaffold Authors
```
```suggestion
Copyright 2019 The Skaffold Authors
```

MarlonGamez commented:
Hey @DanielSel, Thanks for continuously working on this! If you get the chance, would you be able to rebase and fix file conflicts? We should be able to merge this in after that :)

tejal29 commented:
@googlebot I consent

DanielSel commented:
> 
> 
> Hey @DanielSel, Thanks for continuously working on this! If you get the chance, would you be able to rebase and fix file conflicts? We should be able to merge this in after that :)

Done!

nkubala reviewed (approved):
@DanielSel thanks for all your hard work on this PR, and sticking with us to see it through!
